autoPyTorch/evaluation/train_evaluator.py

```diff
@@ -0,0 +1,340 @@
+            opt_fold_weights = [w / sum(opt_fold_weights)
+                                for w in opt_fold_weights]
+
+            # train_losses is a list of dicts. It is
+            # computed using the target metric (self.metric).
+            train_loss = np.average([train_losses[i][str(self.metric[0])]
+                                     for i in range(self.num_folds)],
+                                    weights=train_fold_weights,
+                                    )
+
+            opt_loss = {}
+            # self.logger.debug("OPT LOSSES: {}".format(opt_losses if opt_losses is not None else None))
+            for metric in opt_losses[0].keys():
+                opt_loss[metric] = np.average(
+                    [
+                        opt_losses[i][metric]
+                        for i in range(self.num_folds)
+                    ],
+                    weights=opt_fold_weights,
+                )
+            Y_targets = self.Y_targets
+            Y_train_targets = self.Y_train_targets
+
+            Y_optimization_preds = np.concatenate(
+                [Y_optimization_pred[i] for i in range(self.num_folds)
+                 if Y_optimization_pred[i] is not None])
+            Y_targets = np.concatenate([
+                Y_targets[i] for i in range(self.num_folds)
+                if Y_targets[i] is not None
+            ])
+
+            if self.X_valid is not None:
+                Y_valid_preds = np.array([Y_valid_pred[i]
+                                          for i in range(self.num_folds)
+                                          if Y_valid_pred[i] is not None])
+                # Average the predictions of several models
+                if len(Y_valid_preds.shape) == 3:
+                    Y_valid_preds = np.nanmean(Y_valid_preds, axis=0)
+            else:
+                Y_valid_preds = None
+
+            if self.X_test is not None:
+                Y_test_preds = np.array([Y_test_pred[i]
+                                         for i in range(self.num_folds)
+                                         if Y_test_pred[i] is not None])
+                # Average the predictions of several models
+                if len(Y_test_preds.shape) == 3:
+                    Y_test_preds = np.nanmean(Y_test_preds, axis=0)
+            else:
+                Y_test_preds = None
+
+            self.Y_optimization = Y_targets
+            self.Y_actual_train = Y_train_targets
+
+            self.model = self._get_model()
+
+            status = StatusType.SUCCESS
+            self.logger.debug("In train evaluator fit_predict_and_loss, loss:{}".format(opt_loss))
+            self.finish_up(
+                loss=opt_loss,
+                train_loss=train_loss,
+                opt_pred=Y_optimization_preds,
+                valid_pred=Y_valid_preds,
+                test_pred=Y_test_preds,
+                additional_run_info=additional_run_info,
+                file_output=self.num_folds == 1,
```

Comment:
does this have to be adapted? if not this block is redundant with line 105 and could be placed outside the if else

Reply:
the name of the variables are different and they also contain different info, in holdout, the y_*_pred are predictions on one fold, and in cross val its a concatenation of all the folds. I thought it might be confusing if I try to keep the finish_up outside of the if else

Reply:
Okay, resolving this then